ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -262,105 +221,106 @@ public void shutdown() throws ReplicationException {
+    replicaThreadPoolByDc.computeIfAbsent(datacenter,
+        key -> createThreadPool(datacenter, numOfThreadsInPool, startThread));
+  }
 
-      for (int i = 0; i < replicaThreadCount; i++) {
-        // create the list of nodes for the replica thread
-        Map<DataNodeId, List<RemoteReplicaInfo>> replicasForThread = new HashMap<DataNodeId, List<RemoteReplicaInfo>>();
-        int nodesAssignedToThread = 0;
-        while (nodesAssignedToThread < numberOfNodesPerThread) {
-          DataNodeId dataNodeToReplicate = dataNodeIdIterator.next();
-          replicasForThread.put(dataNodeToReplicate,
-              dataNodeRemoteReplicaInfos.getRemoteReplicaListForDataNode(dataNodeToReplicate));
-          dataNodeIdIterator.remove();
-          nodesAssignedToThread++;
-        }
-        if (remainingNodes > 0) {
-          DataNodeId dataNodeToReplicate = dataNodeIdIterator.next();
-          replicasForThread.put(dataNodeToReplicate,
-              dataNodeRemoteReplicaInfos.getRemoteReplicaListForDataNode(dataNodeToReplicate));
-          dataNodeIdIterator.remove();
-          remainingNodes--;
-        }
-        boolean replicatingOverSsl = sslEnabledDatacenters.contains(datacenter);
-        String threadIdentity =
-            "Replica Thread-" + (dataNodeId.getDatacenterName().equals(datacenter) ? "Intra-" : "Inter") + i
-                + datacenter;
-        try {
-          StoreKeyConverter threadSpecificKeyConverter = storeKeyConverterFactory.getStoreKeyConverter();
-          Transformer threadSpecificTransformer =
-              Utils.getObj(transformerClassName, storeKeyFactory, threadSpecificKeyConverter);
-          ReplicaThread replicaThread =
-              new ReplicaThread(threadIdentity, replicasForThread, factory, clusterMap, correlationIdGenerator,
-                  dataNodeId, connectionPool, replicationConfig, replicationMetrics, notification,
-                  threadSpecificKeyConverter, threadSpecificTransformer, metricRegistry, replicatingOverSsl, datacenter,
-                  responseHandler, SystemTime.getInstance());
-          if (replicaThreadPools.containsKey(datacenter)) {
-            replicaThreadPools.get(datacenter).add(replicaThread);
-          } else {
-            replicaThreadPools.put(datacenter, new ArrayList<>(Arrays.asList(replicaThread)));
-          }
-        } catch (Exception e) {
-          throw new IOException("Encountered exception instantiating ReplicaThread", e);
+  /**
+   * Create thread pool for a datacenter.
+   * @param datacenter The datacenter String.
+   * @param numberOfThreads number of threads to create for the thread pool.
+   * @param startThread If thread needs to be started when create.
+   */
+  private List<ReplicaThread> createThreadPool(String datacenter, int numberOfThreads, boolean startThread) {
+    nextReplicaThreadIndex.put(datacenter, new AtomicInteger(0));
+    List<ReplicaThread> replicaThreads = new ArrayList<>();
+    if (numberOfThreads <= 0) {
```

Comment:
numberOfThreads < 0 should throw IllegalArgumentException

Reply:
This is a bug for ReplicationManger. ReplicationManager should be able to start even if number of threads in remote DCs is 0. But if NPE or other exception is throwed, ReplicationManager constructor failed.

So a if condition is added here, please take a look at.